### PR TITLE
net: lwm2m: Add timeouts to state machine

### PR DESCRIPTION
**Pull request by SeppoTakalo**
Allow certain RD-client states to timeout.

As stated in LwM2M specification:
    The bootstrap procedure failed when the LwM2M Client did not
    receive the "Bootstrap-Finish" operation after the
    EXCHANGE_LIFETIME time period expired.
    The EXCHANGE_LIFETIME parameter is defined in RFC 7252

We must handle the case where Bootstrap server is not sending information towards us.

### CODE REVIEW

**Comment by SeppoTakalo** [comment]
I have an example of error case that this PR will fix:
Screenshot from 2023-09-11 15-21-24

Certain servers (in this case AVSystem Coiote) might block us from starting a new bootstrap until the previous one have finnished, or properly timed out. So I did a trial where I initiated the bootstrap and killed the client immediately. So the bootstrap session is left hanging in the server side.
When the new bootstrap starts, it does not receive any information. Same can happen if Bootsrap-Finnish is lost in the network, but it is much harder to replicate.

If the Bootstrap server accepted our DTLS handshake and accepted our registration message, our client did not have any timeout. It just sits there waiting for packets that are not coming.

**Review by jukkar — APPROVED** [review]
LGTM

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 1]
```diff
@@ -1312,10 +1319,12 @@ static void lwm2m_rd_client_service(struct k_work *work)
 
 		case ENGINE_BOOTSTRAP_REG_SENT:
 			/* wait for bootstrap registration done */
+			timeout = EXCHANGE_LIFETIME;
 			break;
 
 		case ENGINE_BOOTSTRAP_REG_DONE:
 			/* wait for transfer done */
+			timeout = EXCHANGE_LIFETIME;
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:1327]: I see the problem with `ENGINE_BOOTSTRAP_REG_DONE` as in this particular case there is no timeout implemented, as we rely on server-initiated communication (Bootstrap-Finish). But do we need to introduce a separate timeout mechanism for other states? We already have timeout callbacks called when there is no response after serveral retransmission attempts (CoAP-driven):
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_rd_client.c#L442

Comment (SeppoTakalo) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c:1327]: Rest of the timeouts are based on CoAP layer signalling us.

But there is not timeout if there is no CoAP transaction ongoing. Therefore I think the only way to introduce timeouts is to implement this kind of watch-dog mechanism that is not supposed to trigger when all is working correctly.

Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c:1327]: Hmm, I had an impression that being in one of the `_SENT` states implies that there is an ongoing transaction, but maybe there are some cases I don't see. Anyway, having such how you nicely called it "watch-dog" at the state machine level shouldn't hurt anyway.

**Review by rlubos — APPROVED** [review]

**Pull request merged**